enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+---
+title: olm-deletes-operands
+authors:
+  - "@hasbro17"
+reviewers:
+  - "@benluddy"
+  - "@ecordell"
+  - "@kevinrizza"
+  - "@njhale"
+approvers:
+  - "@kevinrizza"
+  - "@dmesser"
+  - "@ecordell"
+  - "@shawn-hurley"  
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional
+see-also:
+  - "N/A"
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# olm-deletes-operands
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The purpose of this enhancement is to outline how the [Operator Lifecycle Manager (OLM)][olm] can let operator users opt-in to the clean up of their operands, or operator workloads, upon the uninstallation of their operator. Specifically when uninstalling an operator, OLM would first delete all CustomResources (CRs) managed by an operator to provide the operator time to clean up all associated resources before deleting the operator itself.
+
+## Motivation
+
+By design when OLM uninstalls an operator it does not remove any of the operator’s owned CustomResourceDefinitions (CRDs) and CRs in order to prevent data loss.
+The operator user or cluster admin is expected to manually delete CRDs, CRs and any other secondary resources created by the operator.
+While some secondary resources that have their [OwnerReference][owner-reference] set to a CR would be garbage collected, others such as off-cluster or cross-namespaced resources with no owner references would remain orphaned.
+
+Without the operator around to process deletion events of CRs, it becomes hard to cleanly remove deployed resources, especially when they exist outside of the cluster.
+
+### Goals
+
+- Clean up includes removal of CRs (operands) and CRDs
+- Operands and CRDs are not automatically removed
+- Upon operator uninstall a user can opt-in to the clean up behavior
+- The operand removal should have a timeout
+- After the operand removal timeout the operator should still be uninstalled
+- Exceeding the operand removal timeout should create an alert event
+
+### Non-Goals
+
+- OLM handles cleanup of anything other than CRs and CRDs
+- Handle custom cleanup logic like filtering/ordering the deletion of CRs 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I can instruct OLM to clean up all of my operator's CRs and CRDs when I uninstall my operator.
+
+#### Story 2
+
+As an operator user I want to see an alert event when OLM does not succeed in operand cleanup after a configured timeout.
```

Comment:
Where deemed appropriate by the author of the operator (e.g. where cleanup does not risk data loss, but lack-of-cleanup results in a mess that's difficult to recover from manually), it would be nice if the operator itself (via the CSV) could indicate that it wants this cleanup to happen by default, unless explicitly switched off by the user.

Reply:
@dmesser @ecordell Do we want to have the same knob in the CSV as well to allow setting the default cleanup behavior for all subscriptions?

That's probably another goal/user-story e.g:
```
As an operator author I would like to configure the default operand cleanup policy for an operator.
```

There is some precedence for having a default in the CSV that is overridden by a Subscription config, e.g CSV specified pod spec's `NodeSelector` vs Subscription config's `NodeSelector`.
https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/subscription-config.md#nodeselector

Reply:
I think this would result in a very fragmented UX that will differ from Operator to Operator. In order to keep the system simple I would not let Operator authors but cluster admins control that at all times, with the same default (do not remove) at all times.

Reply:
Marking as resolved as the CSV is now the point of config. Conversation continued down in https://github.com/operator-framework/enhancements/pull/46#discussion_r511721726